### DISCUSSION TITLE
[JS101] 想請問 Javascript 裡面避免 memory leak 的寫法

### DISCUSSION THREAD

**Discussion by qoosuperman** [Lidemy/forum#54] [前端相關問題]
看了 `從 Object 的等號真正的理解變數` 這堂課，正在思考說有關 memory leak issue 的問題，雖然 Javascript 已經有 garbage collection 的機制，但實務上是不是還有一些細節要特別避免？

看了一些文章，比方說下面這個情況：
`myLeakyFn` 這個 function 的 leak 會變成全域變數，所以 garbage collector 不會自動去回收，造成 memory leak issue
```js
function myLeakyFn() {
  leak = "hello world";
}
```

除了以上這個範例還有什麼寫 javascript 需要注意的地方嗎？

參考的幾篇文章：
https://shawnlin0201.github.io/JavaScript/JavaScript-Garbage-Collection/
https://ithelp.ithome.com.tw/articles/10214185
https://developer.mozilla.org/zh-TW/docs/Web/JavaScript/Memory_Management

**Comment by aszx87410** [comment]
其實一般來說在寫前端的時候會比較少碰到這問題，因為只要頁面重新整理了，記憶體一定會全部都釋放出來，在寫後端的時候因為程式會一直執行，所以會造成比較大的危害。

以前端來說，會碰到的可能就是你問題裡面講的全域變數，或者是鐵人賽文中提到的節點還保存在變數的狀況：

``` js
let a = document.createElement('div');
a.textContent = 'hello world';

document.body.appendChild(a);
document.body.removeChild(a);
console.log(a); // <div>hello world</div>
```

再來就是你附的參考資料也有講到的循環引用，這些都是比較常見的狀況，但其實我覺得在寫 code 的時候滿少會碰到的啦

**Reply by qoosuperman** [reply]
好 感謝~
